### ISSUE TITLE
Strange rendering of toilets in 4.18.0

### ISSUE THREAD

**Issue by giggls** [gravitystorm/openstreetmap-carto#3589]
### Expected behavior
Rendering toilets as in 4.17.0
### Actual behavior
Rendering of toilets is instead very light in 4.18.0
### Links and screenshots illustrating the problem
https://tile.openstreetmap.org/19/274268/181238.png

I have no Idea which causes this. Maybe some opacity setting.

**Comment by Adamant36** [comment]
I assume it has something to do with issue #3528 and it being a private POI. Maybe it needs some slight adjusting. It would help if you provided a link to the exact location of the issue though so people can how its tagged and whatever.

**Comment by Tomasz-W** [comment]
In OSM-Carto, half-transparency of some objects symbolises limeted access to them. 

@giggls Did you open this issue because you didn't know about it at all, you consider it as a bad solution or you think that transaparency is set too light?

**Comment by Tomasz-W** [comment]
I propose to change issue name to "Make private POIs more visible". If there will be more examples of bad visibility of them we could change opacity from `0.33` to `0.5`

**Comment by jragusa** [comment]
> We might considering to change the opacity into a fixed colour value, to avoid the underlying stuff shining through. Similarly, in some issue gravitystorm gave a hint how to paint an underlying colour before applying the transparent feature.

I thought about using an alias for opacity for a moment. It's a rapid fix. You will find below a comparison between current opacity and 0.5 as provided in the former issue:

https://www.openstreetmap.org/way/343862313
`marker-opacity: 0.33;`
![private_shower](https://user-images.githubusercontent.com/9897203/48832582-c4993200-ed79-11e8-8bd4-b4ea4e94e1d8.png)
`marker-opacity: 0.5;`
![private_shower_05](https://user-images.githubusercontent.com/9897203/49298671-439b1280-f4be-11e8-9e3f-3b3073c2d73c.png)

https://www.openstreetmap.org/node/4078433504
`marker-opacity: 0.33;`
![private_waste_basket](https://user-images.githubusercontent.com/9897203/48981348-6602e980-f0d4-11e8-8015-19194c466b1e.png)
`marker-opacity: 0.5;`
![private_waste_basket_05](https://user-images.githubusercontent.com/9897203/49298680-485fc680-f4be-11e8-9eec-730d77c868ed.png)

**Comment by giggls** [comment]
Hm, differences in my case:
marker-opacity: 0.5;
![test-wc-0 5](https://user-images.githubusercontent.com/7996314/50385238-ab660700-06d2-11e9-990b-16f02609d6d2.png)
marker-opacity: 0.33;
![test-wc-0 33](https://user-images.githubusercontent.com/7996314/50385240-b0c35180-06d2-11e9-841e-c5a4096d96df.png)

https://www.openstreetmap.org/way/511308456

I am a bit surprised about double standards here. My pull request about rendering different types of campsites in a different way got rejected as beeing "too specific for a general purpose map".
But what else is this one then?

https://github.com/gravitystorm/openstreetmap-carto/pull/2770

**Comment by kocio-pl** [comment]
This one is about more consistent rendering of private objects (we already did it for parking and playground). Different campsites are also acceptable for me, but IIRC icons were not aligned to our guidelines (bigger than 14 px).

**Comment by giggls** [comment]
OT: I consider it impossible to squeeze a caravan and tent icon into recognizable 14x14 size icons. Thus I might have misinterpreted the reason my patch got rejected.

**Comment by kocio-pl** [comment]
I remember different reasons, it was not only me expressing my view and I don't remember who closed it eventually, so you might be right.

**Comment by Tomasz-W** [comment]
@jragusa Can you provide more test renderings with private POIs with 0.5 opacity?

**Comment by polarbearing** [comment]
Can we test fixed, light colour values as well, to compare that with opacity?

**Comment by jragusa** [comment]
@polarbearing what do you mean by light colour value ? I can made comparisons with 0.33 and 0.5 opacity.

**Comment by polarbearing** [comment]
When I measure the leisure-green I get #0C8416, then with private-opaque 0.5 on the typical background, I get #99D49E, i.e. a very bright green. 

I propose to test fixed values, i.e. non-privat -> #0C8416 and private -> #99D49E, instead of opacity. This would avoid any colour-changing of the transparent icon at the boundary of an object. 

The brighter value could also be derived from the regular one with a function, e.g. the current leisure green is calculated: ```@leisure-green: darken(@park, 60%);```, if there is a darken function there probably is a brighten function.

**Comment by polarbearing** [comment]
Thanks for investigating the lighten function. As fine tuning per value would be required anyway, we could used the fixed ones. I wonder if it really helps us to have calculated colours in the definition block like the 60% darker park. It might be easier to have numeric ones there as well.

**Comment by Tomasz-W** [comment]
Shouldn't it be something like:

`marker-fill:  lighten(@<icon key name>, 50%);`

for all private icons then?

**Comment by meased** [comment]
Here is park background with amenity brown text and amenity brown text lightened 50%:

![lighten](https://user-images.githubusercontent.com/14190496/50922604-c5f0ed80-13ff-11e9-829d-7de33c6a426e.png)

If you lighten brown you just get orange. Dimming is mixing the foreground color with the background color, you have to have knowledge of the background, or you have to use opacity.

**Comment by polarbearing** [comment]
And the mixing is exactly what we should _avoid_, to prevent the effect in the top post. 
If algorithmic lightening does not work, we'd need handpicked values, e.g. 

@amenity-brown: ```#734a08``` -> private ```#B7A38A```
@leisure-green: ```#0C8416``` -> private ```#99D49E```
@transportation-icon: ```#0092da``` -> private ```#9FCFE7```

(Probably the lightened amenity brown just thinned the coffee, while I was putting more milk in it)

**Comment by jeisenbe** [comment]
An option that might work for all colors would be to use `mix(@color-name,
@land-color, 50%);`

**Comment by jeisenbe** [comment]
Here's an examples where I've changed `opacity` to `mix(@color-name,@land-color, 50%);`

Current master (uses opacity)
![access-test-master](https://user-images.githubusercontent.com/42757252/50942523-5432a580-14cc-11e9-9199-e85745731bbc.png)

Mix
![access-test-mix](https://user-images.githubusercontent.com/42757252/50942543-6280c180-14cc-11e9-8dc2-e5b298b32af1.png)

EG changes:
```
-      marker-opacity: 0.33;
+      marker-fill: mix(@leisure-green, @land-color, 50%);
```
and
```
-          text-opacity: 0.33;
+          text-fill: mix(@leisure-green, @land-color, 50%);
```

**Comment by jeisenbe** [comment]
> As for the mix, the result is a solid colour, so what is the advantage
over having the colour defined numerically

Easier to adapt the style to use different colors, and to modify the style
in the future. I suppose this is the same reason we often use
`darken(@color, 10%)` for text and outline colors, rather than defining
each subsidiary color independently.

**Comment by jeisenbe** [comment]
Here's another test with the icons straddling two different background colors:

Current rendering
![opacity-test-current-master](https://user-images.githubusercontent.com/42757252/50978004-cc35b580-1536-11e9-9e3c-e885870646bb.png)

With change to `mix(@<color-name>, @land-color, 50%)`
![mix-test-new](https://user-images.githubusercontent.com/42757252/50978058-ea9bb100-1536-11e9-9144-21d73bd76879.png)

The 50% mix doesn't work for amenity-brown on major buildings (the dark brown background), but the opacity version isn't very visible either. Probably it could be adjusted.

**Comment by Tomasz-W** [comment]
Definetely 'mix' version works better. I wouldn't worry about major buildings, because private POIs inside stations/ churches are propably very rare thing.

**Comment by jeisenbe** [comment]
> are you sure about the rendering of the private label of playground and picnic table ? They look very strong.

No, I'm not sure about those. I just used the same code for all of these examples. As with the brown icons, it would need tuning. Perhaps 40% or 60% would work better?

I hope that it can be improved and perhaps you could do a PR?

**Comment by jeisenbe** [comment]
I've come back to this issue. 

The main problem is that major buildings are very dark. Even some of the standard icons are rather low contrast with the major-building color (eg gastronomy orange, amenity-brown), which limits how much lighter the private icons can become. This has also been mentioned as an issue in #3071 "A main entrance does not stand out well if the building is major" and perhaps somewhere else. 

So I'm submitting a PR to adjust the major building color first: 

65% mix with land-color, with current major buildings color (and several other landcover examples):
![z18-big-private-icons-mix65](https://user-images.githubusercontent.com/42757252/51813697-f63ef400-22fa-11e9-9b72-3e1976ee1b7f.png)

65% mix with land-color, after major-buildings color lightened (to 10% darker than `@building` rather than 20% darker; outlineline is 15% rather than 25%)
![z18-lighter-major-mix65](https://user-images.githubusercontent.com/42757252/51813698-f9d27b00-22fa-11e9-9c0b-9359db750100.png)

**Comment by Adamant36** [comment]
It will also shade out toilets tagged with random access tags, not just private. There was one I came across last-night that was tagged as access=public and was shaded out. So that needs to be looked into.

**Comment by jeisenbe** [comment]
Good point. I agree that it would be better to only change those marked
access=no or =private etc.


It will also shade out toilets tagged with random access tags, not just
> private. There was one I came across last-night that was tagged as
> access=public and was shaded out. So that needs to be looked into.
>

**Comment by jragusa** [comment]
Please note that `access=public` is not listed on the wiki page. According to the wiki, `access=yes` is the correct scheme. 
Otherwise, restricting to `access=no` or `access=private` is the better option to avoid weird behaviour.

**Comment by Adamant36** [comment]
> Please note that access=public is not listed on the wiki page. According to the wiki, access=yes is the correct scheme.

I'm aware that its not a "proper" tag, but either way there shouldn't be conditional rendering that's suppose to be for a specific thing being shown on random tags. It could access=elephant. Its still not in the range of access=private. Which is the what the lightened rendering is suppose to show.

**Comment by HolgerJeromin** [comment]
Fyi
`access=public` is/was a special tag which is defined "toilet only" to make the access clear. It was (is?) included in josm preset, so it should probably supported.

**Comment by Adamant36** [comment]
Interesting. I wasn't aware of that. My bad.
